### ISSUE TITLE
input_device not honored on Windows/dinput

### ISSUE THREAD

**Issue by uli42** [libretro/retroarch-joypad-autoconfig#760]
I have just downloaded and installed RetroArch 1.9.2 on Windows. 

I own a Speedlink Competion Pro USB joystick with Vendor ID 121 and Product ID 6172. For that joystick I have created an autoconfig file that also matches the input_device string (I got that string from the linux dmesg output):
```
input_driver = "dinput"
input_device = "SPEEDLINK COMPETITION PRO Game Controller for Android "
input_device_display_name = "Speedlink Competition Pro Extra SL-650212-BKRD (dinput)"
input_vendor_id = "121"
input_product_id = "6172"
...
```

Although having installed the above autoconfig file my joystick appears as "Beboncool DA04", which stems from the distributed autoconfig file dinput/BBC-GAME.cfg:
````
input_driver = "dinput"
input_device = "BBC-GAME"
input_vendor_id = "121"
input_product_id = "6172"
...
````

If I rename dinput/BBC-GAME.cfg to a name `(e.g. zzBBC_GAME.cfg) that comes alphabetically AFTER my config file the joystick is correctly shown as Speedlink.

As both devices are sharing the same Vendor and Product ID the input_device string is the only factor that can be used to distinguish between them but it seems to be ignored!

Also, I have not found any hints when to use dinput and when to use hid or any of the other input_drivers. Where's that documented?

**Comment by RobLoach** [comment]
It could be that the dinput driver doesn't output information about the device. Think we could just remove BBC-GAME?

**Comment by uli42** [comment]
Yes, of course it can be removed, but the root problem will stay. If dinput does not offer that information the whole concept is problematic and then should use some other input mechanism instead.

**Comment by bslenul** [comment]
A bit late but it seems to work properly for me, I made some tests with a proper "input_device" and an incorrect one, and no matter the .cfg filename it loaded the good one.

> I got that string from the linux dmesg output and it's working there

Maybe that's your issue, for example my 8bitdo controller has a different name in my Windows 10 install (`8Bitdo NES30 Pro`):

![image](https://user-images.githubusercontent.com/33353403/125433519-8545051f-903f-4dcb-8291-2886f28c3aff.png)

and my Linux Mint VM (`8Bitdo NES30 Pro   8Bitdo NES30 Pro`):

![image](https://user-images.githubusercontent.com/33353403/125433482-ff4a9756-ea21-46e7-b992-aefba1e13223.png)

**Comment by uli42** [comment]
I have checked the device name with some USB tools on windows and it matched. It might, however, be the case that dinput uses/composes the name differently. I cannot seem to find helpful documentation for the dinput stuff so may someone with more insight can shine a light on how dinput handles device names? How can I see what dinput _sees_?

**Comment by offalynne** [comment]
On Linux the device name field combines vendor and product names which is why it's in duplicate

**Comment by bslenul** [comment]
Hm you could try USBView: [USBView.zip](https://github.com/libretro/retroarch-joypad-autoconfig/files/6834975/USBView.zip)

Hardest part is to actually find your device in the list :D but then it gives a lot of info, including the name:

![image](https://user-images.githubusercontent.com/33353403/126043352-5632be7d-8497-46f5-a183-2714d494018f.png)

**Comment by uli42** [comment]
I have just tried the USBview from above and it reports this:

```
       ---===>Device Information<===---
English product name: "Game Controller for Android "
...
idVendor:                        0x0079 = Shenzhen Longshengwei Technology, Co., Ltd.
idProduct:                       0x181C
bcdDevice:                       0x0101
iManufacturer:                     0x01
     English (United States)  "SPEEDLINK COMPETITION PRO"
iProduct:                          0x02
     English (United States)  "Game Controller for Android "
iSerialNumber:                     0x00
bNumConfigurations:                0x01
```

So - as expected - the values I see on Windows do not differ from the ones I see in Linux. However, 

Putting it all together it is getting clear that in the dinput code I linked above only the ProductID is being used for the name and not - as I had expected - Manufacturer + ProductID. I have changed my file to ProductID only and now it works:

```
input_driver = "dinput"
input_device = "Game Controller for Android "
input_device_display_name = "Speedlink Competition Pro Extra SL-650212-BKRD (dinput)"
input_vendor_id = "121"
input_product_id = "6172"
input_b_btn = "3"
input_b_btn_label = "Left Fire"
input_a_btn = "0"
input_a_btn_label = "Right Fire"
input_select_btn = "4"
input_select_btn_label = "Left Triangle"
input_start_btn = "1"
input_start_btn_label = "Right Triangle"
input_up_axis = "-1"
input_up_axis_label = "Up"
input_down_axis = "+1"
input_down_axis_label = "Down"
input_left_axis = "-0"
input_left_axis_label = "Left"
input_right_axis = "+0"
input_right_axis_label = "Right"
```

Thanks!

One question remains, though: Why is the manufacturer not used in dinput? Is it not available there?

**Comment by RobLoach** [comment]
Mind sending a Pull Request with the change? I'd love to see more dinput changes get pushed forwards.

**Comment by uli42** [comment]
Yeah, I am going to do that soon.

**Comment by nfp0** [comment]
I seem to be having the same issue, but with a Switch Online Mega Drive/Genesis controller.
It has the same product id and vendor id than the Switch Online SNES controller, but with a different name.
It gets picked up as a SNES controller because of that, so I created a new autoconfig from Retroarch but it gets created as `input_device = "Wireless Gamepad" `
At first sight, this works, but this new autoconfig also picks up the SNES controller, which it shouldn't.
It's as if dinput is always using "Wireless Gamepad" for all Bluetooth controllers.

I also noticed this is what Windows calls these controllers on dinput:
![image](https://user-images.githubusercontent.com/11202422/150135777-a837a6af-ab16-456b-971d-4c2c2da56d00.png)

Even though on the Bluetooth interface they have their proper names:
![image](https://user-images.githubusercontent.com/11202422/150136004-0102a16f-60fa-4c46-995a-cdd0c0cf4c71.png)

On Linux, on udev, everything works fine and the 2 controllers are correctly distinguished through the input_device entry, but not on dinput.
I am going to create a pull request for the udev file for the Switch Online Mega Drive/Genesis controller, but it would be nice to have dinput working too for the Windows folks.

**Comment by uli42** [comment]
Maybe you can find out further details with
https://www.nirsoft.net/utils/bluetooth_viewer.html

**Comment by nfp0** [comment]
@uli42 Here it is, for both controllers:

![image](https://user-images.githubusercontent.com/11202422/150155351-907c370c-ba48-4a05-b488-390ea74a3ea9.png)

![image](https://user-images.githubusercontent.com/11202422/150155436-c71f9e95-e4a0-423f-a0ab-a2dfb1827fe2.png)

It seems to me the "Wireless Gamepad" string is being reported for all Bluetooth controllers on dinput. Even the Switch Pro Controller uses `input_device = "Wireless Gamepad"` when I create a new profile for it in the Retroarch input menu.

**Comment by offalynne** [comment]
Yeah this is definitely an issue but you’ll have to take it up with the parent project; this repo just handles config data submissions. It would definitely be better if DInput Bluetooth devices could correctly ID on name.

**Comment by nfp0** [comment]
Yeah, I'll open an issue there.

**Comment by nfp0** [comment]
Opened an issue here on the RetroArch project:
https://github.com/libretro/RetroArch/issues/13520
